lib/AST/ASTImporter.cpp

```diff
@@ -7947,8 +7982,12 @@ ASTImporter::Import(NestedNameSpecifierLoc FromNNS) {
         return std::move(Err);
       TypeSourceInfo *TSI = getToContext().getTrivialTypeSourceInfo(
             QualType(Spec->getAsType(), 0), ToTLoc);
-      Builder.Extend(getToContext(), ToLocalBeginLoc, TSI->getTypeLoc(),
-                     ToLocalEndLoc);
+      if (Kind == NestedNameSpecifier::TypeSpecWithTemplate)
+        Builder.Extend(getToContext(), ToLocalBeginLoc, TSI->getTypeLoc(),
+                       ToLocalEndLoc);
+      else
+        Builder.Extend(getToContext(), SourceLocation{}, TSI->getTypeLoc(),
```

Comment:
Why is here the begin location an empty SourceLocation? And why it is not in case of `TypeSpecWithTemplate`? What is the difference? `ToLocalBeginLoc` is invalid here?

Reply:
```
  /// Extend the current nested-name-specifier by another
  /// nested-name-specifier component of the form 'type::'.
  ///
  /// \param Context The AST context in which this nested-name-specifier
  /// resides.
  ///
  /// \param TemplateKWLoc The location of the 'template' keyword, if present.
  ///
  /// \param TL The TypeLoc that describes the type preceding the '::'.
  ///
  /// \param ColonColonLoc The location of the trailing '::'.
  void Extend(ASTContext &Context, SourceLocation TemplateKWLoc, TypeLoc TL,
              SourceLocation ColonColonLoc);
```
The second parameter indicates if there is a `template` keyword and this is the case of `TypeSpecWithTemplate`.The `template` is here the begin location of `template X::`, the location of `::` is (always) the end location. The begin location is passed by `TSI->getTypeLoc()` too.

Reply:
Okay, this is clear now, but I think it would worth a short comment at the call site too.